src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1162 @@
+    d_num_completed_md5 = 0;
+
+    memset(d_completed_workers, TOKEN_STATE_PENDING, sizeof(d_completed_workers));
+    memset(d_completed_md5, TOKEN_STATE_PENDING, sizeof(d_completed_md5));
+
+    d_total_ingressed_obj = 0;
+    d_num_failed_workers = 0;
+  }
+
+
+  static constexpr auto COOKIE_LEN = 15;
+  static constexpr auto CLUSTER_ID_LEN = 15;
+  //---------------------------------------------------------------------------
+  cluster::cluster(const DoutPrefixProvider *_dpp,
+                   CephContext *cct,
+                   rgw::sal::Driver* driver):
+    dpp(_dpp),
+    d_lock_cookie(gen_rand_alphanumeric(cct, COOKIE_LEN)),
+    d_cluster_id (gen_rand_alphanumeric(cct, CLUSTER_ID_LEN))
+  {
+    clear();
+    ldpp_dout(dpp, 10) << __func__ << "::cluser_id=" << d_cluster_id << dendl;
+
+    auto store = dynamic_cast<rgw::sal::RadosStore*>(driver);
+    if (!store) {
+      ldpp_dout(dpp, 0) << "ERR: failed dynamic_cast to RadosStore" << dendl;
+      ceph_abort("non-rados backend");
+      return;
+    }
+
+    librados::IoCtx ioctx;
+    if (init_dedup_pool_ioctx(store->getRados(), dpp, ioctx) != 0) {
+      throw std::runtime_error("Failed init_dedup_pool_ioctx()");
+    }
+
+    // generate an empty epoch with zero counters
+    int ret = set_epoch(ioctx, d_cluster_id, dpp, 0, 0);
+    if (ret != 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::failed set_epoch()! ret="
+                        << ret << "::" << cpp_strerror(-ret) << dendl;
+      throw std::runtime_error("Failed set_epoch()");
+    }
+
+    dedup_epoch_t epoch;
+    reset(store, ioctx, &epoch, 0, 0);
+  }
+
+  //---------------------------------------------------------------------------
+  int cluster::reset(rgw::sal::RadosStore *store,
+                     librados::IoCtx &ioctx,
+                     dedup_epoch_t *p_epoch,
+                     work_shard_t num_work_shards,
+                     md5_shard_t num_md5_shards)
+  {
+    ldpp_dout(dpp, 10) << __func__ << "::REQ num_work_shards=" << num_work_shards
+                       << "::num_md5_shards=" << num_md5_shards << dendl;
+    clear();
+
+    while (true) {
+      int ret = get_epoch(ioctx, dpp, p_epoch, __func__);
+      if (ret != 0) {
+        return ret;
+      }
+      if (p_epoch->num_work_shards && p_epoch->num_md5_shards) {
+        ldpp_dout(dpp, 10) << __func__ << "::ACC num_work_shards=" << p_epoch->num_work_shards
+                           << "::num_md5_shards=" << p_epoch->num_md5_shards << dendl;
+        break;
+      }
+      else if (!num_work_shards && !num_md5_shards) {
+        ldpp_dout(dpp, 10) << __func__ << "::Init flow, no need to wait" << dendl;
+        break;
+      }
+      else {
+        ret = swap_epoch(dpp, ioctx, p_epoch,
+                         static_cast<dedup_req_type_t> (p_epoch->dedup_type),
+                         num_work_shards, num_md5_shards);
+      }
+    }
+
+    d_epoch_time = p_epoch->time;
+    cleanup_prev_run(ioctx);
```

Comment:
you ignore any error codes returned from these calls (` cleanup_prev_run()`,  `create_shard_tokens()`).
is it intentional?

Reply:
This code is running in parallel by multiple independent RGW so failures are expected (e.g. 2 RGW members trying to remove the same object)

Reply:
the code of `cleanup_prev_run()` already take care of possible races, by ignoring `-ENOENT` errors.
currently it return an error only if `rgw_list_pool()` return an error different than `-ENOENT`.
if this is happening, it is not due to parallel runs of RGWs.

anyway, if failed cleanup is not a reson to stop the process, you can still ignore a failure in the calling function.
will it create an issue with the current run, or is it just a matter of resource leak?

Reply:
Failing to remove token objects will cause the dedup to skip them as they would appear to be owned by others.
Running a quick pass to verify all tokens were created after EPOCH should prevent this from happening.

SLAB objects are removed when we finish processing them, but if an RGW fails we might be left with lots of SLAB objects consuming space.

Reply:
done